rpc_util.go

```diff
@@ -792,15 +797,25 @@ func decompress(compressor encoding.Compressor, d []byte, maxReceiveMessageSize
 // dc takes precedence over compressor.
 // TODO(dfawley): wrap the old compressor/decompressor using the new API?
 func recv(p *parser, c baseCodec, s *transport.Stream, dc Decompressor, m interface{}, maxReceiveMessageSize int, payInfo *payloadInfo, compressor encoding.Compressor) error {
-	d, err := recvAndDecompress(p, s, dc, maxReceiveMessageSize, payInfo, compressor)
+	buf, err := recvAndDecompress(p, s, dc, maxReceiveMessageSize, payInfo, compressor)
 	if err != nil {
 		return err
 	}
-	if err := c.Unmarshal(d, m); err != nil {
+	if err := c.Unmarshal(buf, m); err != nil {
 		return status.Errorf(codes.Internal, "grpc: failed to unmarshal the received message: %v", err)
 	}
 	if payInfo != nil {
-		payInfo.uncompressedBytes = d
+		if p.sharedRecvBufferPool != nil {
+			if len(buf) != 0 {
+				payInfo.uncompressedBytes = make([]byte, len(buf))
```

Comment:
This seems like the biggest problem with this concept to me.  Doesn't this mean that turning on compression negates any reason to use this feature (and in fact makes it worse)?

Otherwise I think we should really reference count the buffer and only do `Put` once all references are gone.

Reply:
Thankfully, this is not related to compression.

The problem arises only when payInfo != nil. In this case, the value is either logged outside of the function using a `binaryLogger` or stored and passed to statsHandlers as `stats.InPayload.Data`.

We could potentially identify a suitable point to return the value to the pool and apply it, or ask the grpc user to return it.

However, addressing this would deviate significantly from the primary focus of this PR. As such, it would be more appropriate to handle it as a separate issue.